### PR TITLE
bank*: distribute initial balances uniformly

### LINKED ISSUE #94: bank-lua fails with 1.8% probability by construction
The test creates 10 bank accounts and set initial balances:

```
$ grep :read tarantool-3f568c339221ac10948c0dd29e405c0f224597dc/20211029T163440.000Z/jepsen.log | grep -v :invoke | head -n 1
2021-10-29 16:44:43,402{GMT}	INFO	[jepsen worker 0] jepsen.util: 0	:ok	:read	{0 100, 1 0, 2 0, 3 0, 4 0, 5 0, 6 0, 7 0, 8 0, 9 0}
```

The 0th account has balance 100, others have 0.

Next, the test starts to perform transfers from account to account:

```
$ grep :transfer tarantool-3f568c339221ac10948c0dd29e405c0f224597dc/20211029T163440.000Z/jepsen.log | grep -v :invoke | head -n 10
2021-10-29 16:44:44,169{GMT}	INFO	[jepsen worker 0] jepsen.util: 0	:fail	:transfer	{:from 4, :to 2, :amount 5}
2021-10-29 16:44:47,393{GMT}	INFO	[jepsen worker 0] jepsen.util: 0	:fail	:transfer	{:from 9, :to 4, :amount 2}
2021-10-29 16:44:48,177{GMT}	INFO	[jepsen worker 0] jepsen.util: 0	:fail	:transfer	{:from 2, :to 0, :amount 2}
2021-10-29 16:44:49,756{GMT}	INFO	[jepsen worker 0] jepsen.util: 0	:fail	:transfer	{:from 7, :to 6, :amount 5}
2021-10-29 16:44:51,306{GMT}	INFO	[jepsen worker 0] jepsen.util: 0	:fail	:transfer	{:from 4, :to 9, :amount 5}
2021-10-29 16:44:53,736{GMT}	INFO	[jepsen worker 0] jepsen.util: 0	:fail	:transfer	{:from 7, :to 0, :amount 4}
2021-10-29 16:44:54,518{GMT}	INFO	[jepsen worker 0] jepsen.util: 0	:fail	:transfer	{:from 1, :to 2, :amount 5}
2021-10-29 16:44:55,311{GMT}	INFO	[jepsen worker 0] jepsen.util: 0	:fail	:transfer	{:from 2, :to 7, :amount 4}
2021-10-29 16:44:56,089{GMT}	INFO	[jepsen worker 0] jepsen.util: 0	:fail	:transfer	{:from 2, :to 1, :amount 5}
2021-10-29 16:44:57,632{GMT}	INFO	[jepsen worker 0] jepsen.util: 0	:fail	:transfer	{:from 9, :to 3, :amount 4}
```

When 'from' account has not enough balance, the transfer fails (it does not fail the whole test). 'from'/'to' accounts are choosen randomly. When all transfer fail, the test fails (see [here](https://github.com/jepsen-io/jepsen/blob/0.1.18/jepsen/src/jepsen/checker.clj#L169-L186)).

The test is time bounded: 60 seconds. We have no concurrency in the test and amount of transfers is around 38. There is `(9/10)^38 ≈ 1.8%` probability that we'll never attempt to transfer from 0th account, so all transfers will fail.

The result of the test is reported this way in given case:

```
2021-10-29 16:46:06,559{GMT}	INFO	[jepsen test runner] jepsen.core: {:perf
 {:latency-graph {:valid? true},
  :rate-graph {:valid? true},
  :valid? true},
 :clock-skew {:valid? true},
 :crash {:valid? true},
 :timeline {:valid? true},
 :stats
 {:valid? false,
  :count 77,
  :ok-count 39,
  :fail-count 38,
  :info-count 0,
  :by-f
  {:read
   {:valid? true,
    :count 39,
    :ok-count 39,
    :fail-count 0,
    :info-count 0},
   :transfer
   {:valid? false,
    :count 38,
    :ok-count 0,
    :fail-count 38,
    :info-count 0}}},
 :exceptions {:valid? true},
 :workload
 {:SI
  {:valid? true,
   :read-count 39,
   :error-count 0,
   :first-error nil,
   :errors {}},
  :plot {:valid? true},
  :valid? true},
 :valid? false}


Analysis invalid! (ﾉಥ益ಥ）ﾉ ┻━┻
```

(Not quite obvious on the first glance, but look at `:ok-count 0`.)

There are some ways to resolve it or decrease probability of the fail:

* Distribute the initial total amount across all accounts.
* Enable concurrency.
* Always make the first transfer from 0th account.
* Allow negative balances.
* A fix for #96 may also decrease probability of the fail due to insreased overall amount of transfer attempts per second.

Uniform initial distribution or permitting negative balances are only true solutions, they also resolve the problem with a lot of failed transfers. See below.

I looked over several other tests implementations:

* CockroachDB distributed balance uniformly (see [here](https://github.com/jepsen-io/jepsen/blob/0.1.18/cockroachdb/src/jepsen/cockroach/bank.clj#L38-L42) and [here](https://github.com/jepsen-io/jepsen/blob/0.1.18/cockroachdb/src/jepsen/cockroach/bank.clj#L165)).
* Ignite distributes them too (see [here](https://github.com/jepsen-io/jepsen/blob/0.1.18/ignite/src/jepsen/ignite/bank.clj#L71-L76), [here](https://github.com/jepsen-io/jepsen/blob/0.1.18/ignite/src/jepsen/ignite/bank.clj#L18-L19) and [here](https://github.com/jepsen-io/jepsen/blob/0.1.18/ignite/src/jepsen/ignite/bank.clj#L40)).
* YugabyteDB sets the first account to an initial value and others to zero (see [here](https://github.com/jepsen-io/jepsen/blob/0.1.18/yugabyte/src/yugabyte/ycql/bank.clj#L28-L35) and [here](https://github.com/jepsen-io/jepsen/blob/0.1.18/yugabyte/src/yugabyte/ysql/bank.clj#L29-L35)).

So, it seems, there is no canonical way to choose initial balances. OTOH, if we'll distribute them uniformly, we'll make more successful transfers, so the test will be more powerful.

Full log: [logs.txt](https://github.com/tarantool/jepsen.tarantool/files/7446927/logs.txt).
More detailed info: [jepsen-single-instance.zip](https://github.com/tarantool/jepsen.tarantool/files/7446930/jepsen-single-instance.zip) (see `tarantool-3f568c339221ac10948c0dd29e405c0f224597dc/20211029T163440.000Z/`).

----

I would also highlight that even when the test in its current implementation does not fails, it generates much less successful transactions, if we'll compare it with the uniform balances distribution case or negative balances case. So our test lacks some stress testing power just because of the bad initial balances distribution.

### PR DESCRIPTION
**Pull request by Totktonada**
Before this commit the initial distribution was the following: first
account has `total-amount` balances, other 9 ones have zero balance.

From time to time is appears that there are no attempts to transfer from
the first account. It means that all transfers are marked as failed and
the history analysis reports fail.

Even when there are some successful transfers, there are many attempts
to transfer from an account with zero balance: those transfers are
useless in terms of tarantool testing.

This commit changes the initial distribution to the uniform one and
resolves the problems, which are described above. See the linked issue
for details.

Aside of the problematic bank-lua I changed other banking tests in the
same way to unify the approach. How each test is changed:

* bank-lua: [100 0 0 0 0 0 0 0 0 0] -> [10 10 10 10 10 10 10 10 10 10].
* bank-multitable-lua: the test already distributes initial balances
  uniformly, but hardcodes balance value. Now it is deduced from the
  `total-amount` parameter.
* bank: same as bank-lua.
* bank-multitable: same as bank-lua.

'bank' and 'bank-multitable' fixes are blind, because they're disabled.
See #83.

'bank-lua' and 'bank-multitable-lua' are passed in a manual run.
'bank-lua' gives 2-3 failed transfers of ~40 ones. Before it was like
5-15 *successful* ones of 30-40. (I have no representative statistics,
just looked over several runs in CI, but the numbers looks expected.)

Fixes #94

### CODE REVIEW

**Comment by Totktonada** [comment]
@ligurio Please, look if time permits. My reason to fix it is just because I don't want to see those periodical fails.

I tested it manually using CI. The results are [here](https://github.com/tarantool/tarantool/commits/589f8f20ef53c049459e5584300a15542715d56b) (those logs will be expired after some time). However I tried cluster testing as well and persisted logs and artifacts in #98. Don't be confused by another commit hash: the code is the same, I just updated the commit message.

File: src/tarantool/bank.clj [hunk 1]
```diff
@@ -37,10 +37,13 @@
                              balance INT NOT NULL)")])
            (doseq [a (:accounts test)]
                (info "Populating account")
-               (sql/insert! conn table-name {:id      a
-                                             :balance (if (= a (first (:accounts test)))
-                                                       (:total-amount test)
-                                                       0)}))))
+               ; Distribute initial balances uniformly.
+               (let [initial-balance (/ (:total-amount test)
```
Comment (ligurio) [inline src/tarantool/bank.clj]: I would rename `initial-balance` to `initial-balance-per-account`.

File: src/tarantool/bank.clj [hunk 2]
```diff
@@ -37,10 +37,13 @@
                              balance INT NOT NULL)")])
            (doseq [a (:accounts test)]
                (info "Populating account")
-               (sql/insert! conn table-name {:id      a
-                                             :balance (if (= a (first (:accounts test)))
-                                                       (:total-amount test)
-                                                       0)}))))
+               ; Distribute initial balances uniformly.
```
Comment (ligurio) [inline src/tarantool/bank.clj]: We can move `let` construction before `doseq`. It's more than enough to calculate `initial-balance` once.

**Review by ligurio — CHANGES_REQUESTED** [review]
Thanks! I like your idea to distribute balances uniformly.

**Comment by Totktonada** [comment]
Re-verified after review fixes: [results](https://github.com/tarantool/tarantool/commits/6caa95317668ac417fed51650540d03b871674da).

File: src/tarantool/bank.clj (hunk 1 above)
Comment (Totktonada) [reply src/tarantool/bank.clj]: It feels a bit too long for me, but okay. Changed.

File: src/tarantool/bank.clj (hunk 2 above)
Comment (Totktonada) [reply src/tarantool/bank.clj]: Well, I can just add it into `let` above and don't even increase depth level. Done.

File: src/tarantool/bank.clj [hunk 3]
```diff
@@ -27,7 +27,10 @@
 
   (setup! [this test node]
     (locking BankClientWithLua
-      (let [conn (cl/open node test)]
+      (let [conn (cl/open node test)
+            ; Distribute initial balances uniformly.
+            initial-balance-per-account (/ (:total-amount test)
+                                           (count (:accounts test)))]
```
Comment (ligurio) [inline src/tarantool/bank.clj:36]: What will happen when user will pass float `total-amount` or/and `accounts` to test?
What will happen when user will pass an `account` equal to zero?

**Review by ligurio — CHANGES_REQUESTED** [review]

File: src/tarantool/bank.clj (hunk 3 above)
Comment (Totktonada) [reply src/tarantool/bank.clj:36]: > What will happen when user will pass float `total-amount`

Initially I had idea to distribute balances almost uniformly, when it is not possible to do that in a strictly uniform way. Say, for total-balance 10 and three accounts assign balances to `[4 4 2]`.

```lua
local left_amount = total_amount
local usual_balance = math.ceil(total_amount / #accounts)
local balances = {}
for i = 1, #accounts - 1 do
    balances[i] = usual_balance
    left_amount = left_amount - usual_balance
end
balances[#accounts] = left_amount
```

But, to be honest, I don't know Clojure good enough to don't make such code look like mess.

I thought about consequences of just 'balance is total_amount/#accounts` and, I guess, everything should work well: we'll just fill accounts with floating point numbers. Well, I didn't test it, but I doubt that we'll need such parameters in a foreseeable future.

> What will happen when user will pass float <...> `accounts` to test?

Accounts are set as an array, so amount of items cannot be float.

> What will happen when user will pass an `account` equal to zero?

If you meant, an empty array of accounts...

https://github.com/tarantool/tarantool/issues/3077#issuecomment-369734110

Seriously, I'm not very interested about how this code will behave for senseless testing parameters.

Comment (Totktonada) [reply src/tarantool/bank.clj:36]: We discussed it with Sergey and agreed that it worth to add assertion that we'll not fall into floating point numbers.

**Comment by Totktonada** [comment]
Re-verified after review fixes: [results](https://github.com/tarantool/tarantool/commits/239867f00058bbd34ffc3c6be5f9a10eea0965fe).

File: src/tarantool/bank.clj (hunk 3 above)
Comment (Totktonada) [reply src/tarantool/bank.clj:36]: Added assertions.

Comment (ligurio) [reply src/tarantool/bank.clj:36]: Output in test `bank` with default 'total-amount' and 'accounts' values:
```
INFO [2021-11-24 10:25:53,457] jepsen worker 0 - jepsen.core Running worker 0
INFO [2021-11-24 10:25:53,457] jepsen nemesis - jepsen.core Running nemesis
INFO [2021-11-24 10:25:53,475] jepsen worker 0 - jepsen.util 0  :invoke :read   nil
INFO [2021-11-24 10:25:53,589] jepsen worker 0 - jepsen.util 0  :ok     :read   {0 10, 1 10, 2 10, 3 10, 4 10, 5 10, 6 10, 7 10, 8 10, 9 10}
INFO [2021-11-24 10:25:53,589] jepsen worker 0 - jepsen.util 0  :invoke :read   nil
INFO [2021-11-24 10:25:53,665] jepsen worker 0 - jepsen.util 0  :ok     :read   {0 10, 1 10, 2 10, 3 10, 4 10, 5 10, 6 10, 7 10, 8 10, 9 10}
INFO [2021-11-24 10:25:53,665] jepsen worker 0 - jepsen.util 0  :invoke :read   nil
INFO [2021-11-24 10:25:53,753] jepsen worker 0 - jepsen.util 0  :ok     :read   {0 10, 1 10, 2 10, 3 10, 4 10, 5 10, 6 10, 7 10, 8 10, 9 10}
```

With ':total-amount' equal to 101 and default 'accounts' value test failed:
```
INFO [2021-11-24 10:28:05,897] jepsen node 37.139.42.230 - jepsen.core downloading /var/log/tarantool/jepsen.log to jepsen.log
INFO [2021-11-24 10:28:05,939] jepsen node 37.139.42.230 - tarantool.db 37.139.42.230 Stopping Tarantool
WARN [2021-11-24 10:28:06,693] main - jepsen.core Test crashed!                                                                                               
java.lang.AssertionError: Assert failed: Unable to distribute initial balances uniformly for given total-amount and accounts
(integer? initial-balance-per-account)      
```

Comment (Totktonada) [reply src/tarantool/bank.clj:36]: Thanks for verifying it.

**Review by ligurio — APPROVED** [review]

**Pull request merged**